gems/sorbet-runtime/lib/types/private/methods/_methods.rb

```diff
@@ -108,12 +127,32 @@ def self.signature_for_method(method)
 
   # Only public because it needs to get called below inside the replace_method blocks below.
   def self._on_method_added(hook_mod, method_name, is_singleton_method: false)
-    current_declaration = T::Private::DeclState.current.active_declaration
-    return if !current_declaration
-    T::Private::DeclState.current.reset!
-
     mod = is_singleton_method ? hook_mod.singleton_class : hook_mod
     original_method = mod.instance_method(method_name)
+    current_declaration = T::Private::DeclState.current.active_declaration
+
+    if !current_declaration.nil? && current_declaration.is_final?
+      if !mod.is_a?(Class)
+        raise "`#{mod.name}` is not a class and its method `#{method_name}` cannot be declared as final"
+      end
+      # use hook_mod and not mod here, because we don't bother marking the singleton class of a final class as final.
+      if T::Private::Final.is_final?(hook_mod)
```

Comment:
Should the singleton class be declared as final as well? Then we could use `mod` here.